File: elixir/bob/lib/bob.ex
defmodule Bob do
  @moduledoc """
  Implementation of the Bob exercise in Elixir track on Exercism.
  """
  @spec hey(String.t()) :: String.t()
  def hey(input) do
    cond do
      shouting?(input) and question?(input) -> "Calm down, I know what I'm doing!"
      silence?(input) -> "Fine. Be that way!"
      shouting?(input) -> "Whoa, chill out!"
      question?(input) -> "Sure."
      true -> "Whatever."
    end
  end

  defp shouting?(sentence) do
    sentence == String.upcase(sentence) and sentence != String.downcase(sentence)
  end

  defp question?(sentence) do
    sentence
    |> String.trim()
    |> String.ends_with?("?")
  end

  defp silence?(sentence) do
    String.trim(sentence) == ""
  end
end

File: typescript/bob/bob.ts
export function hey(message: string): string {
  if(shouting(message) && question(message)) {
    return "Calm down, I know what I'm doing!"
  } else if(silence(message)) {
    return 'Fine. Be that way!'
  } else if(shouting(message)) {
    return 'Whoa, chill out!'
  } else if(question(message)) {
    return 'Sure.'
  }

  return 'Whatever.'
}

const question = (message: string): boolean => (
  message.trim().slice(-1) === "?"
)

const shouting = (message: string): boolean => (
  message === message.toUpperCase() && message !== message.toLowerCase()
)

const silence = (message: string): boolean => (
  !message.trim()
)
